From 90c9ebdc19a495665bfebc1c70f4d175185c5e01 Mon Sep 17 00:00:00 2001
From: Amelie Delaunay <amelie.delaunay@foss.st.com>
Date: Fri, 4 Apr 2025 10:38:47 +0200
Subject: [PATCH 1689/1695] driver: stm32_dbgmcu_mbx: check if an IT is pending
 on async notif event

Instead of unconditionally executing stm32_dbgmcu_mbx_threaded_itr() which
processes a request from debugger, prior check DBGMCU message acknowledge
authentication register (AUTH_ACK) if debugger has written AUTH_HOST to
request an authenticated debug session.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 3f892eb90bb3f500de5f32cd3c487289dd043e7b driver: stm32_dbgmcu_mbx: check if an IT is pending on async notif event]
Signed-off-by: Amelie Delaunay <amelie.delaunay@foss.st.com>
Change-Id: I5004c95d05525bc40105b37a41885c795f2979cf
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/449142
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/stm32_dbgmcu_mbx.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_dbgmcu_mbx.c b/core/drivers/stm32_dbgmcu_mbx.c
index 722db243f..0a2caa9f6 100644
--- a/core/drivers/stm32_dbgmcu_mbx.c
+++ b/core/drivers/stm32_dbgmcu_mbx.c
@@ -166,10 +166,13 @@ static enum itr_return stm32_dbgmcu_mbx_itr_auth_rd(struct itr_handler *h
 static void yielding_stm32_dbgmcu_mbx_notif(struct notif_driver *ndrv __unused,
 					    enum notif_event ev)
 {
+	vaddr_t reg_auth_ack = dbgmcu_d.base + DBGMCU_DBG_AUTH_ACK;
+
 	switch (ev) {
 	case NOTIF_EVENT_DO_BOTTOM_HALF:
 		DMSG("Notif DO_BOTTOM_HALF");
-		stm32_dbgmcu_mbx_threaded_itr();
+		if (io_read32(reg_auth_ack) & DBGMCU_DBG_AUTH_ACK_HOST)
+			stm32_dbgmcu_mbx_threaded_itr();
 		break;
 	case NOTIF_EVENT_STOPPED:
 		DMSG("Notif STOPPED");
-- 
2.47.2

